1config-ui/src/com/brunobonacci/oneconfig/ui/core.cljs

```diff
@@ -0,0 +1,535 @@
+    [clojure.string :as string]))
+
+
+; for `println` to work
+(def debug?
+  ^boolean js/goog.DEBUG)
+(enable-console-print!)
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Vars
+(defonce state
+         (atom
+           {
+            ;; contains the config entries retrieved from the server
+            :entries []
+            ;; filters
+            :filters {:key "", :env "", :version ""}
+            ;; manages the toggle for the extended mode
+            :extended-mode? false
+            ;; one of :listing, :new-entry-mode, :show-entry-mode
+            :client-mode :listing
+            ;:new-version-flag? nil
+            :1config-version {:current "" :latest ""}
+
+            :new-entry {:key     ""
+                        :env     ""
+                        :version ""
+                        :type    ""
+                        :val     ""
+                        :file     ""
+                        }
+
+            ;; modal window (initial as plain, should be nested) it should be  ":show-entry-mode"
+            :page-key    :surface-13
+            :item-data   nil
+            :item-params nil
+            ;:modal {
+            ;        :page-key    :surface-13
+            ;        :item-data   nil
+            ;        :item-params nil
+            ;        }
+
+                      }))
+
+;TODO maybe all atoms which manage states should be turned into a single "state"-atom
+(defonce sidenav-display-toggle (atom "sidenav hidden"))
+(defonce file-upload-name (atom ""))
+(defonce file-upload-style (atom "hide-element"))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Ajax Handlers
+
+(defn error-handler [{:keys [status status-text]}]
+  (pp/pprint (str "failure occurred: " status " " status-text)))
+
+(defn error-handler-alert [response]
+  (js/alert response))
+
+(defn error-handler-console [response]
+  (print response))
+
+(defn get-item-handler [response]
+  (swap! state assoc-in [:item-data] (get response :value))
+  (swap! state update :page-key
+         (fn [pk]
+           (if (= pk :surface-13-modal)
+             :surface-13
+             :surface-13-modal))))
+
+(defn footer-element [version]
+  [:div {:class "footer" }
+   (str "1Config.A library to manage multiple environments and application configuration safely and effectively.Apache License 2.0. Bruno Bonacci, 2019, v." (get version :current)
+        (if (= (get version :current) (get version :latest))
+          ""
+          (str "(Latest version v." (get version :latest) ")")))
+   ])
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Ajax
+
+(defn get-all-configs! []
+  (GET "/configs"
+       {
+        :handler         #(swap! state assoc-in [:entries] %)
+        :format          :json
+        :response-format :json
+        :keywords?       true
+        :error-handler   error-handler}))
+
+(defn get-config-item! [item]
+  (let [{:keys [key env version change-num content-type]} item
+        get-url (str "/configs/keys/" key "/envs/" env "/versions/" version )]
```

Comment:
use `format` is more readable than `str` when there are many parts to concatenate

Reply:
Additionally the path parameters need to be url encoded